### DISCUSSION TITLE
Neutral Stations

### DISCUSSION THREAD

**Discussion by 2TallTyler** [OpenTTD/OpenTTD#13065] [Features]
The following is a feature proposal for comments (and so I don't lose my notes 😛). It's written like a PR but no code exists yet.

## Problem/Motivation

Players in multiplayer games want to collaborate and connect their networks, particularly when building passenger networks with cargodist.

The existing infrastructure sharing patch is complex and in my opinion, not suited for inclusion in vanilla OpenTTD.

## Description

Introduce neutral, town-owned stations buildable by players.

To construct a neutral station, hold Ctrl when placing to open the Join menu. In addition to existing station and "New Station", there is a "New Station (neutral)" option.

Neutral stations follow the pattern of Oil Rig stations: they have grey station signs and can be used by any player. Any company-colour pixels in the station sprites are also grey to match.

To expand, modify, remove tiles from, or delete a neutral station, either:
- The company must have a station rating at the station, indicating that they have used it previously, or
- The station must have no station ratings from other companies

This provides a modicum of protection for rail-only stations, although stations that include road, ship, or aircraft facilities can easily be accessed by a bad actor, giving them a station rating and thus modification/demolition rights. The intended usecase for neutral station is in private multiplayer games with trusted players.

Construction costs of neutral stations are charged to the company making the changes as if it were their own station, but the town pays infrastructure maintenance costs. This makes airports much cheaper, at the cost of any player being able to use it. We might consider a setting to enable/disable neutral stations, or maybe just enable/disable neutral airports.

## Limitations

Multi-company deliveries are not paid out properly: only the last company gets paid. This should be fixed in a separate PR.

JGRPP solves this in two commits, but it uses a new 3rd party library which is not present in vanilla OpenTTD.
* Fix: https://github.com/JGRennison/OpenTTD-patches/commit/9000d74f3c80edb9d325bb6980f9391b14c9ab93
* Cache validation: https://github.com/JGRennison/OpenTTD-patches/commit/49dee941fb63f758d61380cedb91953aadc6c1b9
* Library: https://github.com/JGRennison/OpenTTD-patches/commit/b4cfb1adbcfb65d5a10ab18e596046d009cf93a7

**Comment by PeterN** [comment]
This can't be ported straight from JGRPP.

1) JGRPP has a bizarro version of saveload that bypasses description tables, often just serializing bytes as is.
2) The btree can be substituted with a std:: container of some sort, depending on how it's used. (Performance differs of course)

But I think the concept is strange.

Road vehicles, ships and aircraft are probably fine, they are all self contained and vehicles of different companies can already travel on (drive-through) road stops and docking locations.

Rail stations though? That would need some special handling...

**Comment by telk5093** [comment]
As a user who uses JGRPP's infra share feature, this suggestion don't satisfy me since neither it can connect between two companies' railways directly without a (neutral) station within, nor many companies (more than two or three) can use a one railway line, since those are why I use the infra share feature.
And I agree with like PeterN said about road vehciles, ships and aircrafts.

**Comment by wensimehrp** [comment]
This idea only allows the exchange of cargo between different networks. Currently, vehicles cannot run on rival tracks. Before this is changed I don't think neutral (by the way maybe "public" is a better name?) stations would be as powerful.

**Reply by nielsmh** [reply]
Having trains run on other company track comes with a whole lot of extra complexity, in the form of usage fees tracking. It also makes the question of how to handle the griefing opportunities even more important.

**Reply by TheMowgliMan** [reply]
> Having trains run on other company track comes with a whole lot of extra complexity, in the form of usage fees tracking. It also makes the question of how to handle the griefing opportunities even more important.

I'm not sure that has to be done. Usage fees can be left to a GS as with road vehicles (This might be a good reason to allow multiple GSes in a game, but I digress).

**Comment by 2TallTyler** [comment]
I have refined and expanded this idea, in #13498.
